config/configManager.go

```diff
@@ -0,0 +1,195 @@
+
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "failed to load policy provider from config")
+	}
+
+	logrus.Infof("policies successfully created.")
+
+	return stores, verifiers, policyEnforcer, nil
+}
+
+func reloadExecutor(configFilePath string, executor *ef.Executor) error {
+	cf, err := Load(configFilePath)
+
+	if err != nil {
+		logrus.Warnf("failed to load from config file , err: %v", err)
+		return err
+	}
+
+	stores, verifiers, policyEnforcer, err := createFromConfig(cf)
+
+	if err != nil {
+		logrus.Warnf("failed to store/verifier/policy objects from config, no updates loaded. err: %v", err)
+		return err
+	}
+
+	if configHash != cf.FileHash {
+		executor.ReloadAll(stores, verifiers, policyEnforcer, &cf.ExecutorConfig)
+		configHash = cf.FileHash
+		logrus.Infof("configuration file has been updated, reloading executor succeeded")
+	} else {
+		logrus.Infof("no change found in config file, no executor update needed")
+	}
+	return nil
+}
+
+// Setup a watcher on file at configFilePath, reload executor on file change
+func watchForConfigurationChange(configFilePath string, executor *ef.Executor) error {
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		errors.Wrap(err, "new file watcher on configuration file failed ")
+	}
+
+	// setup for loop to listen for events
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					logrus.Warnf("no longer watching configuration file changes, file watcher event channel closed")
+					return
+				}
+
+				logrus.Infof("file watcher event detected %v", event)
+
+				// In a cluster scenario, a configMap will recreate the config file
+				// after the remove event, the watcher will also be removed
+				// since a watcher on a non existent file is not supported, we sleep until the file exist add the watcher back
+				if event.Name == configFilePath && event.Op&fsnotify.Remove == fsnotify.Remove {
+					logrus.Infof("config remove event detected")
+
+					_, err := os.Stat(configFilePath)
+					for err != nil {
+						logrus.Infof("config file does not exist yet, sleeping again")
+						_, err = os.Stat(configFilePath)
+						time.Sleep(1 * time.Second)
+					}
+
+					err = watcher.Add(configFilePath)
+
+					if err != nil {
+						logrus.Errorf("adding configuration file watcher failed, err: %v", err)
+						continue
+					}
+
+					logrus.Infof("watcher added on configuration directory %v", configFilePath)
+					reloadExecutor(configFilePath, executor)
+				}
+
+				// In a local scenario, the configuration will be updated through a write event
+				if event.Name == configFilePath && event.Op&fsnotify.Write == fsnotify.Write {
+					reloadExecutor(configFilePath, executor)
+				}
+
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					logrus.Errorf("configuration file watcher returned error : %v, watcher will be closed.", err)
+					return
+				}
+			}
+		}
+	}()
+
+	err = watcher.Add(configFilePath)
```

Comment:
I am curious if the above go routine addition can be avoided if the watcher add fails? So is there a benefit of moving the go routine after this line 185?

Reply:
routine moved. thanks